syncstorage/views/__init__.py

```diff
@@ -319,6 +342,84 @@ def post_collection(request):
     return res
 
 
+def post_collection_transaction(request):
+    storage = request.validated["storage"]
+    userid = request.validated["userid"]
+    collection = request.validated["collection"]
+    bsos = request.validated["bsos"]
+    invalid_bsos = request.validated["invalid_bsos"]
+    batch = request.validated["batch"]
+    commit = request.validated["commit"]
+
+    # The "batch" key is set only on a multi-POST batch request prior to a
+    # commit.  The "modified" key is only set upon a successful commit.
+    # The two flags are mutually exclusive.
+    # Any failures at all mean cancelling the transaction completely.
+    res = {'success': [], 'failed': {}}
+
+    # If there are any parsing failures, we won't even start a transaction.
```

Comment:
This is an aspect of the Sync protocol that I kinda hate: Sync essentially operates like `INSERT OR IGNORE`, including for malformed input, which is so rarely what one wants!

Fortunately, in the case of batched uploads, the client can see that some of the uploaded BSOs were rejected, and can choose to not commit the entire transaction. 

I think it's fine to keep the existing behavior (queue the ones that were fine) for that reason; if anything, the sane thing to do the reverse: to reject part-wrong non-batch `POST`s and part-accept batched ones!

Two observations:

- It'd be nice to have a `?discard` operation, just for efficiency's sake.
- Careful clients should always issue an empty `POST` to commit, so that they don't accidentally commit a partial final batch.